aspnetcore/security/authorization/secure-data/samples/final3/Pages/Contacts/Details2.cshtml.cs

```diff
@@ -0,0 +1,56 @@
+using ContactManager.Authorization;
+using ContactManager.Data;
+using ContactManager.Models;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using System.Threading.Tasks;
+
+namespace ContactManager.Pages.Contacts
+{
+    [AllowAnonymous]
+    public class Details2Model : DI_BasePageModel
+    {
+        public Details2Model(
+            ApplicationDbContext context,
+            IAuthorizationService authorizationService,
+            UserManager<IdentityUser> userManager)
+            : base(context, authorizationService, userManager)
+        {
+        }
+
+        public Contact Contact { get; set; }
+
+        #region snippet
+        public async Task<IActionResult> OnGetAsync(int id)
+        {
+            Contact = await Context.Contact.FirstOrDefaultAsync(m => m.ContactId == id);
+
+            if (Contact == null)
+            {
+                return NotFound();
+            }
+
+            if (!User.Identity.IsAuthenticated)
+            {
+                return Challenge();
+            }
+
+            var isAuthorized = User.IsInRole(Constants.ContactManagersRole) ||
+                               User.IsInRole(Constants.ContactAdministratorsRole);
+
+            var currentUserId = UserManager.GetUserId(User);
+
+            if (!isAuthorized
+                && currentUserId != Contact.OwnerID
+                && Contact.Status != ContactStatus.Approved)
+            {
+                return Forbid();
+            }
+
+            return Page();
+        }
```

Comment:
The sample code from the linked issue was re "documents" rather than "contacts", so I've copied `Details` from this sample and updated it to `[AllowAnonymous]` and then added the `Challenge` and `Forbid` distinction using the authz logic that was there before.

Reply:
This works for demo purposes, but I think it's a bit unusual that we have authz handlers for everything except reads, which means the authz logic is all inline here. I'd expect additional handlers to take care of the role checks and the approved status, which would be more inline with the rest of the app. That would also allow the following in the page:

```cs
if (!User.Identity.IsAuthenticated)
{
    return Challenge();
}

var authorizationResult = await AuthorizationService
    .AuthorizeAsync(User, Contact, ContactOpertations.Read);

if (!authorizationResult.Succeeded)
{
    return Forbid();
}

return Page();
```